src/server/worker/pipeline/transform/registry.go

```diff
@@ -0,0 +1,951 @@
+	if err != nil {
+		return err
+	}
+
+	// Initialize the datumsBase value in the registry
+	if parentCommitInfo != nil {
+		var err error
+		reg.datumsBase, err = reg.getDatumSet(parentCommitInfo.Datums)
+		if err != nil {
+			return err
+		}
+	} else {
+		reg.datumsBase = make(map[string]struct{})
+	}
+
+	return nil
+}
+
+func isDatumNew(hash string, parentJobs []*pendingJob, datumsBase map[string]struct{}) bool {
+	for i := len(parentJobs) - 1; i >= 0; i-- {
+		if _, ok := parentJobs[i].datumsAdded[hash]; ok {
+			return false
+		}
+		if parentJobs[i].orphan {
+			// No other jobs matter since this job doesn't depend on any previous
+			// hashtree, so we can shortcut out
+			return true
+		}
+		if _, ok := parentJobs[i].datumsRemoved[hash]; ok {
+			return true
+		}
+	}
+
+	if _, ok := datumsBase[hash]; ok {
+		return false
+	}
+	return true
+}
+
+// calculateRemovedDatums walks back through the sets of added and removed
+// datums in upstream pending jobs and determine which datums were removed for
+// the latest pending job. This is done so we can properly reprocess a datum if
+// it is removed then readded.
+func calculateRemovedDatums(
+	allDatums map[string]struct{},
+	parentJobs []*pendingJob,
+	datumsBase map[string]struct{},
+) map[string]struct{} {
+	removed := make(map[string]struct{})
+	skip := make(map[string]struct{})
+
+	for i := len(parentJobs) - 1; i >= 0; i-- {
+		for hash := range parentJobs[i].datumsAdded {
+			if _, ok := skip[hash]; !ok {
+				skip[hash] = struct{}{}
+				if _, ok := allDatums[hash]; !ok {
+					removed[hash] = struct{}{}
+				}
+			}
+		}
+		if parentJobs[i].orphan {
+			// No other jobs matter since this job doesn't depend on any previous
+			// hashtree, so we can shortcut out
+			return removed
+		}
+		for hash := range parentJobs[i].datumsRemoved {
+			skip[hash] = struct{}{}
+		}
+	}
+
+	for hash := range datumsBase {
+		if _, ok := skip[hash]; !ok {
+			skip[hash] = struct{}{}
+			if _, ok := allDatums[hash]; !ok {
+				removed[hash] = struct{}{}
+			}
+		}
+	}
+
+	return removed
+}
+
+// Based on the current state of the registry and pending jobs, calculate the
+// set of new and removed datums in the datum iterator. Also add the input
+// metadata for new datums to the metadata commit.
+func calculateDatumSets(
+	pipelineInfo *pps.PipelineInfo,
+	dit datum.Iterator,
+	parentJobs []*pendingJob,
+	datumsBase map[string]struct{},
+) (datumsAdded map[string]struct{}, datumsRemoved map[string]struct{}, orphan bool) {
+	allDatums := make(map[string]struct{})
+	datumsAdded = make(map[string]struct{})
+
+	// Iterate over datum iterator, comparing to upstream jobs datum sets to find
+	// new datums
+	dit.Reset()
+	for dit.Next() {
+		hash := common.HashDatum(pipelineInfo.Pipeline.Name, pipelineInfo.Salt, dit.Datum())
+
+		allDatums[hash] = struct{}{}
+
+		// Check if the hash was added or removed in any upstream jobs
+		if isDatumNew(hash, parentJobs, datumsBase) {
```

Comment:
I think you can calculate the added and removed datums in one place instead of calculating added in this loop and removed in a separate function. I think it would be much cleaner to just get all of the datums in this loop, then calculate the added / removed in a function with the same parameters as `calculateRemovedDatums` (and similar structure of walking back through the pending jobs, then base) and then compute whether the job is an orphan. Also, small nit would be that I think the name should be `ancestorJobs` (or maybe just `pendingJobs` since that is the type) not `parentJobs`. I am just as guilty of this because I call the base hash tree `parentHashtree`.